src/GraphQL.MemoryCache/MemoryDocumentCache.cs

```diff
@@ -0,0 +1,71 @@
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+
+namespace GraphQL.Caching
+{
+    /// <summary>
+    /// A basic implementation of a document cache, limited by a configured amount of memory.
+    /// </summary>
+    public class MemoryDocumentCache : IDocumentCache, IDisposable
+    {
+        private readonly IMemoryCache _memoryCache;
+        private readonly MemoryDocumentCacheOptions _options;
+
+        /// <summary>
+        /// Initializes a new instance with the default options: 100,000 maximum total query size and no expiration time.
+        /// </summary>
+        public MemoryDocumentCache()
+            : this(new MemoryDocumentCacheOptions())
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance with the specified options.
+        /// </summary>
+        /// <param name="options">A value containing the <see cref="MemoryDocumentCacheOptions"/> to use.</param>
+        public MemoryDocumentCache(IOptions<MemoryDocumentCacheOptions> options)
+            : this(
+                new MemoryCache(new MemoryCacheOptions { SizeLimit = options.Value.MaxTotalQueryLength }),
+                options)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance with the specified memory cache and sliding expiration time period.
+        /// Note that by overriding <see cref="GetMemoryCacheEntryOptions(string)"/>, the sliding expiration
+        /// time specified within <paramref name="options"/> can be ignored.
+        /// </summary>
+        protected MemoryDocumentCache(IMemoryCache memoryCache, IOptions<MemoryDocumentCacheOptions> options)
+        {
+            _memoryCache = memoryCache ?? throw new ArgumentNullException(nameof(memoryCache));
+            _options = options?.Value;
+        }
+
+        /// <summary>
+        /// Returns a <see cref="MemoryCacheEntryOptions"/> instance for the specified query.
+        /// Defaults to setting the <see cref="MemoryCacheEntryOptions.SlidingExpiration"/> value as specified
+        /// in the constructor, and the <see cref="MemoryCacheEntryOptions.Size"/> value to the length of the query.
+        /// </summary>
+        protected virtual MemoryCacheEntryOptions GetMemoryCacheEntryOptions(string query)
+        {
+            return new MemoryCacheEntryOptions { SlidingExpiration = _options.SlidingExpiration, Size = query.Length };
+        }
+
+        /// <inheritdoc/>
+        public virtual Document this[string query]
+        {
+            get => _memoryCache.TryGetValue<Document>(query, out var value) ? value : null;
+            set => _memoryCache.Set(query ?? throw new ArgumentNullException(nameof(query)), value, GetMemoryCacheEntryOptions(query));
+        }
+
+        /// <summary>
+        /// Disposes of the underlying <see cref="IMemoryCache"/> instance.
+        /// </summary>
+        public virtual void Dispose()
+        {
+            _memoryCache.Dispose();
```

Comment:
```suggestion
            if (_owned)
                _memoryCache.Dispose();
```

```c#
private readonly bool _owned;
```
\+ set `_owned` to `true/false` in appropriate ctors.

Reply:
This is not typical for .NET.  For example, `StreamWriter`'s `Dispose` method disposes of it's underlying `Stream` regardless of whether it was constructed with a `Stream` or a `string`.  If the user does not want the underlying object disposed, they should not call `Dispose` on this object.

Reply:
(I don't disagree with your logic, but it doesn't match the .NET standard.)

Reply:
>  If the user does not want the underlying object disposed, they should not call Dispose on this object.

😕 This question is outside the scope of his desires when it comes to work within the DI-container. Imagine that `DocumentCache` is configured like scoped/transient. Then after some request `DocumentCache.Dispose()` kills `IMemoryCache` dependency configured as singleton and oooooops.

Reply:
I understand the problems with that design in those circumstances (typically the scope would be the same, of course).  I am stating that it is not the .NET standard that programmers are accustomed to.  We have to deviate off from the standard if we implement the change.

Reply:
Example with `StreamWriter` is bad example. The base rule is `DO NOT DISPOSE OBJECTS WHICH YOU DIDN'T CREATE` and it works in the vast majority of cases, especially in the realities of modern .NET Core apps, when 99% of apps use DI.

Reply:
Do you think this change is worth breaking the well-developed .NET standard for how Dispose works here?

Reply:
Are you saying that the example is bad or `StreamWriter` is bad?  I use `StreamWriter` all the time.  And I agree that it's a flawed design.

Reply:
If you say that we should break off from the standard, I'll make the change.  I don't have a problem changing it.

Reply:
Since the constructor in question is protected, it is very likely that it IS owned.  So I will add an argument to the constructor as well.

Reply:
>  well-developed .NET standard for how Dispose works

Reply:
However `StreamWriter` was written. I do not know if it is documented.  It just "is".

Reply:
> Are you saying that the example is bad or StreamWriter is bad? 

`StreamWriter` design _was_ bad. So one day MS team added one more overload to mitigate this https://stackoverflow.com/questions/2666888/is-there-any-way-to-close-a-streamwriter-without-closing-its-basestream

And since `StreamWriter` is bad then example based on it is bad too...

Reply:
Never said it wasn't.

Reply:
Then I don't understand the point of giving a bad design counter-argument that itself uses bad design. Well, or I translated something wrong.